doc/rationale.md

```diff
@@ -0,0 +1,84 @@
+# Configy rationale
+
+The following explains the rationale versus various decisions made in Configy.
+
+## Use cases, Internal vs External schema
+
+Configy was originally designed with an application-first mindset: a developer should
+design their config file based on a `struct` definition they chose, and would potentially
+have a slightly different configuration file depending on the problem domain (for example,
+the `@Key("name")` attribute).
+
+While this is still the primary use-case for Configy, it is also very convenient to
+write a validator for schema that are outside of the application's control,
```

Comment:
This is a little confusing to be because writing a validator for a schema seems orthogonal to being able to read schemas outside the application's control.

For validation, probably the best tool is a schema definition such as https://json-schema.org/.

Reply:
I have used it on a few occasions to quickly (as in, fast development time) validate that the config file passed is valid (I can give you links on Slack). I have been very pleased with the experience, but it might be a case of me only having ~Configy~ a hammer :)
We tried json-schema recently and it was very verbose, so we ended up using the library for that (but json-schema would have otherwise been my goto).

Reply:
> Validating a configuration file (and discarding the parsed values) seems to be one use case

But that any JSON / YAML parser can do, Configy is only about reading & validating. IMO it doesn't make sense to separate deserialization and validation, as the two are usually entangled (e.g. you need to validate anything that will affect parsing, such as the enumerated values in the `fromYAML` example).

Reply:
First, I think it does make sense to separate them, as these are two different use cases. For example a validator could focus on providing good error messages and support additional validation semantics (i.e. validating domain-specific types), while a deserializer only cares about the data. Configy does both, but that's not true for all similar projects, so it's good to be aware of the distinction in the documentation.

Second... I'm not sure where this is going. If you say "we need X for validation" but then explain that X is actually used not just for validation but also deserialization I think that's confusing, because deserialization necessarily involves some degree of validation anyway.

Reply:
I think I might be overthinking this. Would the following work in your opinion ?
```suggestion
read and validate a schema that are outside of the application's control,
```

Reply:
Yep!